curve25519-dalek/src/elligator2.rs

```diff
@@ -0,0 +1,786 @@
+    /// RFC.
+    ///
+    /// The input r and output P are elements of the field F. Note that
+    /// the output P is a point on the edwards curve and as such it's byte
+    /// representation is distinguishable from uniform random.
+    ///
+    /// Input:
+    ///     * r -> an element of field F.
+    ///
+    /// Output:
+    ///     * P - a point on the Edwards elliptic curve.
+    ///
+    /// See <https://datatracker.ietf.org/doc/rfc9380/>
+    pub fn map_to_point(r: &[u8; 32]) -> EdwardsPoint {
+        let mut clamped = *r;
+        clamped[31] &= MASK_UNSET_BYTE;
+        let r_0 = FieldElement::from_bytes(&clamped);
+        let (x, y) = map_fe_to_edwards(&r_0);
+        Self::from_xy(&x, &y)
+    }
+
+    #[cfg(feature = "elligator2")]
+    fn from_xy(x: &FieldElement, y: &FieldElement) -> EdwardsPoint {
+        let z = FieldElement::ONE;
+        let t = x * y;
+
+        EdwardsPoint {
+            X: *x,
+            Y: *y,
+            Z: z,
+            T: t,
+        }
+    }
+
+    /// Maps a representative to a curve point.
+    ///
+    /// This function is the inverse of `to_representative`.
+    pub fn from_representative<T: MapToPointVariant>(representative: &[u8; 32]) -> Option<Self> {
+        T::from_representative(representative).into()
+    }
+
+    /// Mapts a point on the curve to a representative.
+    pub fn to_representative<T: MapToPointVariant>(&self, tweak: u8) -> Option<[u8; 32]> {
+        T::to_representative(&self.to_montgomery().0, tweak).into()
+    }
+}
+
+// ===========================================================================
+// Randomized implementation
+// ===========================================================================
+
+/// Gets a public representative for a key pair using the private key (RFC 9380).
+///
+/// The `tweak` parameter is used to adjust the computed representative making
+/// it computationally indistinguishable from uniform random. If this property
+/// is not required then the provided tweak value does not matter.
+///
+/// The tweak allows us to overcome three limitations:
+/// - Representatives are not always canonical.
+/// - Bit 255 (the most significant bit) was always zero.
+/// - Only points from the large prime-order subgroup are represented.
+///
+/// In order for the returned representative to be canonical a tweak to the
+/// high order two bits must be applied.
+/// ```txt
+/// [An adversary could] observe a representative, interpret it as a field
+/// element, square it, then take the square root using the same
+/// non-canonical square root algorithm. With representatives produced by
+/// an affected version of [the elligator2 implementation], the output of
+/// the square-then-root operation would always match the input. With
+/// random strings, the output would match only half the time.
+/// ```
+///
+/// For a more in-depth explanation see:
+/// <https://github.com/agl/ed25519/issues/27>
+/// <https://www.bamsoftware.com/papers/fep-flaws/>
```

Comment:
I don't understand the distinction. The two things you linked above do not seem to point out any flaw in the RFC, rather implementation errors in some libraries. Is there something inherent in the RFC that requires modification? The impl above already appears to do randomization.

It appears that the actual issue is that the above RFC 9380 mapping always clears the cofactor, which is a leaky thing to do. If that's the issue, why not change the API to do the real mappings point -> field rather than scalar -> point -> field?